hip/components/cooperative_groups.hip.hpp

```diff
@@ -0,0 +1,445 @@
+    detail::is_synchronizable_group_impl<xstd::decay_t<T>>;
+
+
+/**
+ * Check if T is a CommunicatorGroup.
+ */
+template <typename T>
+using is_communicator_group =
+    detail::is_communicator_group_impl<xstd::decay_t<T>>;
+
+
+// types
+
+
+namespace detail {
+
+
+/**
+ * This is a limited implementation of the HIP thread_block_tile.
+ * `any` and `all` are only supported when the size is config::warp_size
+ *
+ */
+template <size_type Size>
+class thread_block_tile {
+public:
+    __device__ thread_block_tile()
+        : data_{Size,
+                static_cast<unsigned>(
+                    (threadIdx.x +
+                     blockDim.x * (threadIdx.y + blockDim.y * threadIdx.z)) %
+                    Size)}
+    {}
+
+    __device__ __forceinline__ unsigned thread_rank() const noexcept
+    {
+        return data_.rank;
+    }
+
+    __device__ __forceinline__ unsigned size() const noexcept
+    {
+        return data_.size;
+    }
+
+    __device__ __forceinline__ void sync() const noexcept {}
+
+#define GKO_BIND_SHFL(ShflOp, ValueType, SelectorType)                       \
+    __device__ __forceinline__ ValueType ShflOp(                             \
+        ValueType var, SelectorType selector) const noexcept                 \
+    {                                                                        \
+        return __##ShflOp(var, selector, Size);                              \
+    }                                                                        \
+    static_assert(true,                                                      \
+                  "This assert is used to counter the false positive extra " \
+                  "semi-colon warnings")
+
+    GKO_BIND_SHFL(shfl, int32, int32);
+    GKO_BIND_SHFL(shfl, float, int32);
+    GKO_BIND_SHFL(shfl, uint32, int32);
+
+    GKO_BIND_SHFL(shfl_up, int32, uint32);
+    GKO_BIND_SHFL(shfl_up, uint32, uint32);
+    GKO_BIND_SHFL(shfl_up, float, uint32);
+
+    GKO_BIND_SHFL(shfl_down, int32, uint32);
+    GKO_BIND_SHFL(shfl_down, uint32, uint32);
+    GKO_BIND_SHFL(shfl_down, float, uint32);
+
+    GKO_BIND_SHFL(shfl_xor, int32, int32);
+    GKO_BIND_SHFL(shfl_xor, float, int32);
+    GKO_BIND_SHFL(shfl_xor, uint32, int32);
+
+    __device__ __forceinline__ int any(int predicate) const noexcept
+    {
+        static_assert(Size == kernels::hip::config::warp_size,
+                      "Hip does not have subwarp any.");
+        return __any(predicate);
+    }
+
+    __device__ __forceinline__ int all(int predicate) const noexcept
+    {
+        static_assert(Size == kernels::hip::config::warp_size,
+                      "Hip does not have subwarp all.");
+        return __all(predicate);
+    }
+
+    __device__ __forceinline__ uint64_t ballot(int predicate) const noexcept
+    {
+        static_assert(Size == kernels::hip::config::warp_size,
+                      "Hip does not have subwarp ballot.");
+        return __ballot(predicate);
+    }
```

Comment:
We only have power-of-two thread block tiles, right? Could we use a bit shift and subwarp-bitmask to implement this for smaller tile sizes?

Reply:
I am not sure whether I understanding it correctly.
I can implement `any`, `all`, `ballot` by reduction, but I need to check whether I can call reduce in this structure or I need implement a simple verion reduce.

Reply:
Something like `(__ballot(predicate) >> (threadIdx.x / subwarp_size * subwarp_size)) & (subwarp_size - 1)` for a power-of-two `subwarp_size` computes the ballot result only for a subwarp. We can then compare this result to 0 or `subwarp_size - 1` to get `any` and `all`.

Reply:
It should work when all threads of warp at the same operation.
Does it work when subwarps are not in the same `if-else`? (some in `if` other in `else`)
By the way, ginkgo does not need this subwarp `any`, `all`, `ballot` currently, so I will implement it in this PR.
If we implement these operation, we also need tests to make sure it works properly.

Reply:
That depends on your execution model, I guess. If you have independent thread scheduling like on the Volta GPU, this would surely not work. But if all threads execute the same instruction (possibly predicated if we are in another branch of the `if-else`), I see no reason why it shouldn't, since we mask the results from the other subwarps. But you are correct, this is outside the scope of this PR.